lib/tapioca/dsl/compilers/active_record_associations.rb

```diff
@@ -292,6 +306,29 @@ def type_for(reflection)
           T.must(qualified_name_of(reflection.klass))
         end
 
+        sig do
+          params(
+            reflection: ReflectionType,
+          ).returns(String)
+        end
+        def single_association_type_for(reflection)
+          return "T.untyped" if column_type_option.untyped?
+
+          association_class = type_for(reflection)
+          return as_nilable_type(association_class) unless column_type_option.persisted?
+
+          case reflection
+          when ActiveRecord::Reflection::BelongsToReflection
+            return association_class if reflection.options[:optional].nil? || !reflection.options[:optional]
```

Comment:
I was confusing so noting that `nil?` looks at more than expected https://github.com/rails/rails/blob/ac104520ee8518441bb4f8c2c7835d0bb8fd6800/activerecord/lib/active_record/relation/query_attribute.rb#L35-L40

Reply:
Hm, apologies but I don't follow your comment. I added the `nil?` check because often times these options are not passed to the association. In that case, we should fall back to default behavior. Let me know if I'm missing something.